lib/manifest.go

```diff
@@ -0,0 +1,381 @@
+	}
+	defer func() {
+		if err1 := a.unlock(); err == nil {
+			err = err1
+		}
+	}()
+	return a.man.SetUser(user)
+}
+func (a *ACBuild) SetWorkingDir(dir string) (err error) {
+	if err = a.lock(); err != nil {
+		return err
+	}
+	defer func() {
+		if err1 := a.unlock(); err == nil {
+			err = err1
+		}
+	}()
+	return a.man.SetWorkingDir(dir)
+}
+func (a *ACBuild) SetTag(tag string) (err error) {
+	if err = a.lock(); err != nil {
+		return err
+	}
+	defer func() {
+		if err1 := a.unlock(); err == nil {
+			err = err1
+		}
+	}()
+	return a.man.SetTag(tag)
+}
+
+func (a *ACBuild) AddDependency(imageName types.ACIdentifier, imageId *types.Hash, labels types.Labels, size uint) (err error) {
+	if err = a.lock(); err != nil {
+		return err
+	}
+	defer func() {
+		if err1 := a.unlock(); err == nil {
+			err = err1
+		}
+	}()
+	switch m := a.man.(type) {
+	case *appc.Manifest:
+		return m.AddDependency(imageName, imageId, labels, size)
+	}
+	return fmt.Errorf("dependencies only supported in appc builds")
+}
+func (a *ACBuild) RemoveDependency(imageName string) (err error) {
+	if err = a.lock(); err != nil {
+		return err
+	}
+	defer func() {
+		if err1 := a.unlock(); err == nil {
+			err = err1
+		}
+	}()
+	switch m := a.man.(type) {
+	case *appc.Manifest:
+		return m.RemoveDependency(imageName)
+	}
+	return fmt.Errorf("dependencies only supported in appc builds")
+}
+func (a *ACBuild) AddIsolator(name string, value []byte) (err error) {
+	if err = a.lock(); err != nil {
+		return err
+	}
+	defer func() {
+		if err1 := a.unlock(); err == nil {
+			err = err1
+		}
+	}()
+	switch m := a.man.(type) {
+	case *appc.Manifest:
+		return m.AddIsolator(name, value)
+	}
+	return fmt.Errorf("dependencies only supported in appc builds")
+}
+func (a *ACBuild) RemoveIsolator(imageName string) (err error) {
+	if err = a.lock(); err != nil {
+		return err
+	}
+	defer func() {
+		if err1 := a.unlock(); err == nil {
+			err = err1
+		}
+	}()
+	switch m := a.man.(type) {
+	case *appc.Manifest:
+		return m.RemoveIsolator(imageName)
+	}
+	return fmt.Errorf("dependencies only supported in appc builds")
+}
+func (a *ACBuild) SetName(name string) (err error) {
+	if err = a.lock(); err != nil {
+		return err
+	}
+	defer func() {
+		if err1 := a.unlock(); err == nil {
+			err = err1
+		}
+	}()
+	switch m := a.man.(type) {
+	case *appc.Manifest:
+		return m.SetName(name)
+	}
+	return fmt.Errorf("setting names only supported in appc builds")
+}
+func (a *ACBuild) SetPostStop(exec []string) (err error) {
+	if err = a.lock(); err != nil {
+		return err
+	}
+	defer func() {
+		if err1 := a.unlock(); err == nil {
+			err = err1
+		}
+	}()
+	switch m := a.man.(type) {
+	case *appc.Manifest:
+		return m.SetPostStop(exec)
+	}
+	return fmt.Errorf("event handlers only supported in appc builds")
+}
+func (a *ACBuild) SetPreStart(exec []string) (err error) {
+	if err = a.lock(); err != nil {
+		return err
+	}
+	defer func() {
+		if err1 := a.unlock(); err == nil {
+			err = err1
+		}
+	}()
```

Comment:
I think I couldn't find a way to when I first wrote this since the deferred closure is using the fact that the named return variable is in scope to be able to check if the function returned with an error, and pulling that out into a named function would lose the closure properties, but I'll take another look.